### ISSUE TITLE
NullPointer Exception when starting embedded elastic 

### ISSUE THREAD

**Issue by roytmana** [elastic/elasticsearch#16810]
**Elasticsearch version**: 2.2

**JVM version**: 1.7

**OS version**:Win10

**Description of the problem including expected versus actual behavior**:

I use embedded (servlet) elastic via transport wares when moving to 2.2 I started getting exception on startup. they do not preven elastic from functioning but I wanted to let you know anyways.
I have just one embedded node

I have 0 replicas and I have 

`discovery.zen.ping.multicast.enabled: false`

it could be something specific to my workstation because explicetly putting my IPv4 address instead default loopback addresses into `discovery.zen.ping.unicast.hosts` fixes it but I decided to report NPE anyway so you can handle it better perhaps.

```text
INFO  02/25/16 13:02:36 elasticsearch.transport - [es_alexr_ems_n1] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9300}, {127.0.0.1:9300}
INFO  02/25/16 13:02:36 elasticsearch.discovery - [es_alexr_ems_n1] es_alexr_ems/EcvRVvceQea3Evecm7O5Fg
WARN  02/24/16 22:11:24 transport.netty - [es_alexr_ems_n1] exception caught on transport layer [[id: 0x710c9066, /0:0:0:0:0:0:0:1:59963 => /0:0:0:0:0:0:0:1:9301]], closing connection
java.lang.NullPointerException
	at org.elasticsearch.transport.netty.MessageChannelHandler.handleException(MessageChannelHandler.java:206)
	at org.elasticsearch.transport.netty.MessageChannelHandler.handlerResponseError(MessageChannelHandler.java:201)
	at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:136)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
WARN  02/24/16 22:11:24 transport.netty - [es_alexr_ems_n1] exception caught on transport layer [[id: 0xa927d869, /127.0.0.1:59967 => /127.0.0.1:9301]], closing connection
java.lang.NullPointerException
	at org.elasticsearch.transport.netty.MessageChannelHandler.handleException(MessageChannelHandler.java:206)
	at org.elasticsearch.transport.netty.MessageChannelHandler.handlerResponseError(MessageChannelHandler.java:201)
	at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:136)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)

```

**Comment by eskibars** [comment]
@roytmana you're likely interested in https://github.com/elastic/elasticsearch/issues/12721 .  The transport-wares plugin is aged and we're not seeing a lot of real usage of it out in the community

**Comment by roytmana** [comment]
@eskibars Yeah I seen it and I am not happy it is being discontinued as there are enough people using it. I had to make some changes in the code for it to work with 2.2

I do not think the is related to transport-wares I am pretty sure I will get the same if running standalone ES. I think it is some connectivity issue on my workstation that's causing it and wanted the development team to see the exception because NPE is never a good thing

**Comment by jasontedor** [comment]
@roytmana Can you provide a minimal reproduction *without* transport wares? I am not able to reproduce on master nor 2.2 currently.

**Comment by roytmana** [comment]
@jasontedor it does not need any special reproduction. I just unzipped standard elastic distro and it started it with the same exceptions. I think it is something specific to my workstation when connecting to from localhost to localhost (maybe from ipv6 to ipv4) but in any case elastich should not fail with NPE but handle it better. Attached is log with TRACE level 

[elasticsearch.log.txt](https://github.com/elastic/elasticsearch/files/147275/elasticsearch.log.txt)

**Comment by jasontedor** [comment]
@roytmana Can you give any more details about your system and config? I'm also running IPv4 and IPv6 and I do not see this issue. While avoiding the NPE is easy here, I want to understand the root cause before addressing.

**Comment by roytmana** [comment]
I wish I could - I am not a networking guru but I think it is the same
issue I had with some development tool that start jetty on localhost and
then tries to connect to it. It fails. I had to force to use my real ipv4
address to work. I think it started couple of months ago with one windows
routine updates and I think my co-worked is having the same issue with that
tool after that update. It is only loopback connections that seem to be
failing. If you can tell me what to check I will gladly do. I have already
tried   JDK 7, 8, 32 and 64 bit - no difference I tried to turn ipv6 off
but it does not affect loopback it still resolves to ::1 and did not fix
the problem ...

**Comment by roytmana** [comment]
here seems to be similar problem http://superuser.com/questions/962293/cant-connect-to-socket-listening-on-127-0-0-1-but-works-for-localhost-in-wind but no solution

**Comment by eskibars** [comment]
@roytmana just FYI, I just tried it on my win10 machine as well (with all the latest MS updates and both ipv6/ipv4) and am not getting this.  Can you give some more detail about your networking setup?

**Comment by roytmana** [comment]
@eskibars it could be the antivirus/local firewall. but I stopped it it did not help. I spent half a couple of months ago trying to figure why service listening on loopback can't be connected to on my workstation (the tool I was using could not be configured on which address to listen and to which to connect) but could not figure it out. Hopefully it is not a widespread issue since your workstation worked fine and my notebook too.

I could send you my ipconfig and anything else you may need but I would prefer to do it privately
please let me know what else besides ipconfig /all you would want and where to send it

For my purposes I just set my node to local mode since it will be a single node embedded deployment and the issue went away of course

**Comment by jasontedor** [comment]
I also attempted to reproduce this on a Windows 10 system with the latest updates and it did not reproduce for me. Again, it's really easy to make this NPE go away (without even understanding where it's coming from) but I really want to understand where it's coming from in case it's indicative of a bigger issue and making the NPE go away would just be papering over it.

**Comment by roytmana** [comment]
Once again let me know what info to collect and where to send it

**Comment by jasontedor** [comment]
> Once again let me know what info to collect and where to send it

@roytmana If I made available a custom build of Elasticsearch (or pointed you to a branch that you could build from source (and so see the diffs from v2.2.0)), would you be able to run it and send me the logs?

**Comment by roytmana** [comment]
Yes sure please send me the link to download it

**Comment by jasontedor** [comment]
> Yes sure please send me the link to download it

@roytmana Thanks. I've created a branch [issue-16810](https://github.com/jasontedor/elasticsearch/tree/issue-16810) in my repository that you can [diff](https://github.com/jasontedor/elasticsearch/commit/d1755be587b6c59d6531cc19d647a6411c1f6ebf) against the [v2.2.0 tag](https://github.com/elastic/elasticsearch/tree/v2.2.0) in the elastic repository. 

The salient changes are some extra logging statements that I hope will give us enough information to track down the root cause of the problem.

I've [uploaded](https://drive.google.com/file/d/0B4tiSXOP8wheN2xVY1g3Slg2a3M/view?usp=sharing) a packaged zip distribution of Elasticsearch from commit [d1755be](https://github.com/jasontedor/elasticsearch/commit/d1755be587b6c59d6531cc19d647a6411c1f6ebf) in that issue-16810 branch. It has the following SHA-256 hash:

```
d76fe7251b48644fa741b5a3682e0b02b56c539a77fc1fc8984e496d0cb315dc
```

When you start Elasticsearch from the distribution in this zip, it will print a log line that matches

```
^.*version\[2.2.1-SNAPSHOT\], pid\[[0-9]+\], build\[d1755be/2016-02-27T16:30:37Z\]$
```

Can you run Elasticsearch from this distribution with `TRACE` logging enabled, and after the NPE reproduces, share the output here?

**Comment by roytmana** [comment]
please see attached. it is a zip file but I had to rename it because for some reason github did not want to accept zip file

[elasticsearch.zip.txt](https://github.com/elastic/elasticsearch/files/149530/elasticsearch.zip.txt)

**Comment by jasontedor** [comment]
@roytmana I think that you have a 1.x node running on your system and that's what is going on here. Can you confirm?

**Comment by roytmana** [comment]
@jasontedor not at the same time and in any case all my embedded elastic instances have unique cluster name.
I am pretty sure no 1.x was running at the time I did the test but 2.2 could have. I will double-check everything again make sure no elastic is running and run it again

**Comment by jasontedor** [comment]
> @jasontedor not at the same time and in any case all my embedded elastic instances have unique cluster name.

They still attempt to discover each other, which combined with incompatible serialization is exactly what I think is going on here.

> I am pretty sure no 1.x was running at the time I did the test but 2.2 could have. I will double-check everything again make sure no elastic is running and run it again

I'm fairly convinced this situation can only happen if a 1.x node was running, not if another 2.x node was running, because I think the root cause is incompatible serialization.

**Comment by roytmana** [comment]
I stopped every JVM on my computer and no elastic of any version is running on my computer or network. same issue. logs attached
[elasticsearch.zip.txt](https://github.com/elastic/elasticsearch/files/151320/elasticsearch.zip.txt)

**Comment by jasontedor** [comment]
There is no NullPointerException in the logs that you attached, just `ConnectTransportException`s which are completely expected here. Now that you're sure that there are no other JVMs running on your system, this only confirms my theory even more. :)

```bash
11:53:09 2d [jason:~/Downloads] $ cat elasticsearch.log | grep NullPointerException
11:53:18 2d [jason:~/Downloads] 1 $ 
```

**Comment by roytmana** [comment]
but why all the exceptions connection?
let me start another 2.2 instance and run it again see if NPEs would come back

**Comment by jasontedor** [comment]
> but why all the exceptions connection?

Because when using the default configuration, Elasticsearch tries to discover nodes also bound to localhost on ports 9301--9304. Since you don't have any other instances running, those discovery attempts will hit connection exceptions.

**Comment by roytmana** [comment]
No NPEs with 2.2 started 1.7.2 and NPEs are back. I guess you would need to look into it. Two separate clusters should not interfere with each other.

Why do you says connection exceptions are expected? They go away if I specify non loopback address for my node

**Comment by jasontedor** [comment]
> No NPEs with 2.2 started 1.7.2 and NPEs are back.  

I consider this confirmation of my theory. :)

> I guess you would need to look into it. Two separate clusters should not interfere with each other.

They are not interfering with each other, the `NullPointerException` here ends up being harmless. And this will eventually be addressed by #15971 but there are no plans to make any changes to 1.x and 2.x along these lines.

> Why do you says connection exceptions are expected? They go away if I specify non loopback address for my node

Because the attempted discovery only occurs if you are using the *default network configuration*.
